src/zope/tales/tests/test_expressions.py

```diff
@@ -134,67 +191,67 @@ class AnOldStyleClass:
-        self.assertEqual(expr(context), 'A2')
+        self._check_evals_to(expr, 'A2')
 
     def testStringSub_w_python(self):
         CompilerError = self.engine.getCompilerError()
         self.assertRaises(CompilerError,
                           self.engine.compile,
-                                'string:${python:1}')
+                          'string:${python:1}')
 
     def testStringSubComplex(self):
         expr = self.engine.compile('string:a ${x/y} b ${y/z} c')
-        context=self.context
-        self.assertEqual(expr(context), 'a yikes b 3 c')
+        self._check_evals_to(expr, 'a yikes b 3 c')
 
     def testStringSubComplex_w_miss_and_python(self):
         # See https://bugs.launchpad.net/zope.tales/+bug/1002242
         CompilerError = self.engine.getCompilerError()
         self.assertRaises(CompilerError,
                           self.engine.compile,
-                            'string:${nothig/nothing|python:1}')
+                          'string:${nothig/nothing|python:1}')
 
     def testString8Bits(self):
         # Simple eight bit string interpolation should just work.
-        from six import u
-        if PY3:
-            # Py3: We simply do not handle 8-bit strings.
-            return
+        # Except on Py3, where we really mess it up.
         expr = self.engine.compile('string:a ${eightBits}')
-        context=self.context
-        self.assertEqual(expr(context), u('a déjà vu').encode('utf8'))
+        expected = 'a ' + self.context.vars['eightBits'] if not PY3 else self.py3BrokenEightBits
+        self._check_evals_to(expr, expected)
 
     def testStringUnicode(self):
         # Unicode string expressions should return unicode strings
-        from six import u
-        expr = self.engine.compile(u('string:Fred'))
-        context=self.context
+        expr = self.engine.compile(u'string:Fred')
+        context = self.context
         result = expr(context)
-        self.assertEqual(result, u('Fred'))
-        self.assertTrue(isinstance(result, unicode))
+        self.assertEqual(result, u'Fred')
+        self.assertIsInstance(result, text_type)
 
     def testStringFailureWhenMixed(self):
         # Mixed Unicode and 8bit string interpolation fails with a
         # UnicodeDecodeError, assuming there is no default encoding
-        from six import u
-        if PY3:
-            # Py3: We simply do not handle 8-bit strings.
-            return
-        expr = self.engine.compile(u('string:a ${eightBits}'))
-        self.assertRaises(UnicodeDecodeError, expr, self.context)
+        expr = self.engine.compile(u'string:a ${eightBits}')
+        with self.assertRaises(UnicodeDecodeError):
+            result = expr(self.context)
+            # If we get here, we're on Python 3, which handles this
+            # poorly.
+            self.assertTrue(PY3)
+            self.assertEqual(result, self.py3BrokenEightBits)
+            self.context.vars['eightBits'].decode('ascii') # raise UnicodeDecodeError
```

Comment:
I'm not entirely sure we want to codify this behavior on Python 3 -- it smells like a bug in the Python 3 port that nobody reported yet.

I may be mistaken.

Reply:
The previous version of this test simply skipped:
```python
if PY3:
           # Py3: We simply do not handle 8-bit strings.
           return
```

Given that all strings are already Unicode on Python 3, I couldn't come up with a better way to handle it...

But more basically, this was already the behaviour if you tried this on Python 3; this test just ensures that if we change that behaviour, we know it and it's on purpose.